File: src/org/elixir_lang/heex/element_type/TemplateData.java
package org.elixir_lang.heex.element_type;

import com.intellij.lang.Language;
import com.intellij.lexer.Lexer;
import com.intellij.psi.templateLanguages.TemplateDataElementType;
import com.intellij.psi.templateLanguages.TemplateLanguageFileViewProvider;
import org.jetbrains.annotations.NotNull;

import static org.elixir_lang.heex.lexer.TemplateData.HEEX;
import static org.elixir_lang.heex.psi.Types.DATA;

public class TemplateData extends TemplateDataElementType {
    private final Language templateFileLanguage;

    public TemplateData(@NotNull Language templateFileLanguage) {
        super(
                "HEEX_TEMPLATE_DATA",
                org.elixir_lang.heex.Language.INSTANCE,
                DATA,
                HEEX
        );
        this.templateFileLanguage = templateFileLanguage;
    }

    @NotNull
    @Override
    protected Lexer createBaseLexer(@NotNull TemplateLanguageFileViewProvider templateLanguageFileViewProvider) {
        return new org.elixir_lang.heex.lexer.TemplateData();
    }

    @NotNull
    @Override
    protected Language getTemplateFileLanguage(TemplateLanguageFileViewProvider templateLanguageFileViewProvider) {
        return templateFileLanguage;
    }
}

File: src/org/elixir_lang/heex/lexer/TemplateData.kt
package org.elixir_lang.heex.lexer

import com.intellij.lexer.Lexer
import com.intellij.lexer.MergingLexerAdapterBase
import com.intellij.psi.tree.IElementType
import com.intellij.psi.tree.OuterLanguageElementType
import org.elixir_lang.heex.Language
import org.elixir_lang.heex.lexer.LookAhead
import org.elixir_lang.heex.psi.Types

/**
 * Merges together all HEEx opening, body, and closing tokens into a single HEEx type
 */
class TemplateData : MergingLexerAdapterBase(LookAhead()) {
    private val mergeFunction: MergeFunction = MergeFunction()

    override fun getMergeFunction(): com.intellij.lexer.MergeFunction = mergeFunction

    private inner class MergeFunction : com.intellij.lexer.MergeFunction {
        override fun merge(type: IElementType, originalLexer: Lexer): IElementType = if (type !== Types.DATA) {
            while (true) {
                val originalTokenType = originalLexer.tokenType

                if (originalTokenType != null && originalTokenType !== Types.DATA) {
                    originalLexer.advance()
                } else {
                    break
                }
            }

            HEEX
        } else {
            type
        }
    }

    companion object {
        @JvmField
        val HEEX: IElementType = OuterLanguageElementType("HEEx", Language.INSTANCE)
    }
}